File: mvplibrary/src/main/java/com/junwu/mvplibrary/http/LoggerInterceptor.java
package com.junwu.mvplibrary.http;

import android.support.annotation.Nullable;

import com.orhanobut.logger.Logger;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.nio.charset.Charset;

import okhttp3.Interceptor;
import okhttp3.MediaType;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.ResponseBody;
import okio.Buffer;
import okio.BufferedSource;

/**
 * ===============================
 * 描    述：OkHttp请求和结果打印
 * 作    者：
 * 创建日期：
 * ===============================
 */
public class LoggerInterceptor implements Interceptor {

    public LoggerInterceptor() {

    }

    @Override
    public Response intercept(Chain chain) throws IOException {
        Request request = chain.request();
        //打印请求前的log
        logRequest(request);
        Response originalResponse;
        try {
            originalResponse = chain.proceed(request);
        } catch (Exception exception) {
            throw exception;
        }
        //打印响应结果
        printResult(request, originalResponse);
        return originalResponse;
    }

    /**
     * 打印响应结果
     *
     * @param request
     * @param originalResponse
     * @return
     * @throws IOException
     */
    @Nullable
    private String printResult(Request request, Response originalResponse) throws IOException {
        ResponseBody responseBody = originalResponse.body();
        String bodyString = null;
        boolean hasRequestBody = request.body() != null;
        Buffer requestbuffer = new Buffer();
        if (hasRequestBody) {
            request.body().writeTo(requestbuffer);
        }

        if (isParseable(responseBody)) {
            BufferedSource source = responseBody.source();
            source.request(Long.MAX_VALUE);
            Buffer buffer = source.buffer();
            String encoding = originalResponse.headers().get("Content-Encoding");
            Buffer clone = buffer.clone();
            bodyString = parseContent(responseBody, encoding, clone);

            Logger.json("{\"Url\":\"" + request.url().toString() + "\","
                    + "\"Method\":\"" + request.method() + "\","
                    + "\"Params\":\"" + (hasRequestBody ? parseParams(request.body(), requestbuffer).replaceAll("&", "  ") : "Null") + "\","
                    + "\"Headers\":\"" + request.headers().toString() + "\","
                    + "\"Response\":" + (isJson(responseBody) ? bodyString + "}" : bodyString + "\"}"));
        } else {
            Logger.e("Url: " + request.url().toString() + "\n"
                    + "Method: " + request.method() + "\n"
                    + "Params: " + (hasRequestBody ? parseParams(request.body(), requestbuffer).replaceAll("&", "  ") : "null") + "\n"
                    + "Headers: " + request.headers().toString()
                    + "Error: " + ((responseBody.contentLength() == 0) ? "ResponseBody ContentLength is Zero" : ""));
        }
        return bodyString;
    }

    private void logRequest(Request request) throws IOException {
        Buffer requestbuffer = new Buffer();
        boolean hasRequestBody = request.body() != null;
        if (hasRequestBody) {
            request.body().writeTo(requestbuffer);
        }
        Logger.e("Url: " + request.url().toString() + "\n"
                + "Method: " + request.method() + "\n"
                + "Params: " + (hasRequestBody ? parseParams(request.body(), requestbuffer).replaceAll("&", "  ") : "null") + "\n"
                + "Headers: " + request.headers().toString());
    }

    /**
     * 解析服务器响应的内容
     *
     * @param responseBody
     * @param encoding
     * @param clone
     * @return
     */
    private String parseContent(ResponseBody responseBody, String encoding, Buffer clone) {
        Charset charset = Charset.forName("UTF-8");
        MediaType contentType = responseBody.contentType();
        if (contentType != null) {
            charset = contentType.charset(charset);
        }
        if (encoding != null && encoding.equalsIgnoreCase("gzip")) {//content使用gzip压缩
            return ZipUtil.decompressForGzip(clone.readByteArray(), convertCharset(charset));//解压
        } else if (encoding != null && encoding.equalsIgnoreCase("zlib")) {//content使用zlib压缩
            return ZipUtil.decompressToStringForZlib(clone.readByteArray(), convertCharset(charset));//解压
        } else {//content没有被压缩
            return clone.readString(charset);
        }
    }

    private String parseParams(RequestBody body, Buffer requestbuffer) throws UnsupportedEncodingException {
        if (body.contentType() == null) return "Unknown";
        if (!body.contentType().toString().contains("multipart")) {
            Charset charset = Charset.forName("UTF-8");
            MediaType contentType = body.contentType();
            if (contentType != null) {
                charset = contentType.charset(charset);
            }
            return URLDecoder.decode(requestbuffer.readString(charset), convertCharset(charset));
        }
        return "This Params isn't Text";
    }

    private boolean isParseable(ResponseBody responseBody) {
        if (responseBody.contentLength() == 0) return false;
        return responseBody.contentType().toString().contains("text") || isJson(responseBody);
    }

    private boolean isJson(ResponseBody responseBody) {
        return responseBody.contentType().toString().contains("json");
    }

    private String convertCharset(Charset charset) {
        String s = charset.toString();
        int i = s.indexOf("[");
        if (i == -1)
            return s;
        return s.substring(i + 1, s.length() - 1);
    }
}

File: ktmvplibrary/src/main/java/com/junwu/ktmvplibrary/http/LoggerInterceptor.kt
package com.junwu.ktmvplibrary.http

import com.orhanobut.logger.Logger
import okhttp3.*
import okio.Buffer
import java.io.IOException
import java.io.UnsupportedEncodingException
import java.net.URLDecoder
import java.nio.charset.Charset

/**
 * ===============================
 * 描    述：
 * 作    者：pjw
 * 创建日期：2017/9/28 10:55
 * ===============================
 */
class LoggerInterceptor : Interceptor {

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        val request = chain.request()
        //打印请求前的log
        logRequest(request)
        val originalResponse: Response
        try {
            originalResponse = chain.proceed(request)
        } catch (exception: Exception) {
            throw exception
        }

        //打印响应结果
        printResult(request, originalResponse)
        return originalResponse
    }

    /**
     * 打印响应结果
     *
     * @param request
     * @param originalResponse
     * @return
     * @throws IOException
     */
    @Throws(IOException::class)
    private fun printResult(request: Request, originalResponse: Response): String? {
        val responseBody = originalResponse.body()
        var bodyString: String? = null
        val hasRequestBody = request.body() != null
        val requestbuffer = Buffer()
        if (hasRequestBody) {
            request.body().writeTo(requestbuffer)
        }

        if (isParseable(responseBody)) {
            val source = responseBody.source()
            source.request(java.lang.Long.MAX_VALUE)
            val buffer = source.buffer()
            val encoding = originalResponse.headers().get("Content-Encoding")
            val clone = buffer.clone()
            bodyString = parseContent(responseBody, encoding, clone)

            Logger.json("{\"Url\":\"" + request.url().toString() + "\","
                    + "\"Method\":\"" + request.method() + "\","
                    + "\"Params\":\"" + (if (hasRequestBody) parseParams(request.body(), requestbuffer).replace("&".toRegex(), "  ") else "Null") + "\","
                    + "\"Headers\":\"" + request.headers().toString() + "\","
                    + "\"Response\":" + if (isJson(responseBody)) bodyString!! + "}" else bodyString!! + "\"}")
        } else {
            Logger.e("Url: " + request.url().toString() + "\n"
                    + "Method: " + request.method() + "\n"
                    + "Params: " + (if (hasRequestBody) parseParams(request.body(), requestbuffer).replace("&".toRegex(), "  ") else "null") + "\n"
                    + "Headers: " + request.headers().toString()
                    + "Error: " + if (responseBody.contentLength() == 0L) "ResponseBody ContentLength is Zero" else "")
        }
        return bodyString
    }

    @Throws(IOException::class)
    private fun logRequest(request: Request) {
        val requestbuffer = Buffer()
        val hasRequestBody = request.body() != null
        if (hasRequestBody) {
            request.body().writeTo(requestbuffer)
        }
        Logger.e("Url: " + request.url().toString() + "\n"
                + "Method: " + request.method() + "\n"
                + "Params: " + (if (hasRequestBody) parseParams(request.body(), requestbuffer).replace("&".toRegex(), "  ") else "null") + "\n"
                + "Headers: " + request.headers().toString())
    }

    /**
     * 解析服务器响应的内容
     *
     * @param responseBody
     * @param encoding
     * @param clone
     * @return
     */
    private fun parseContent(responseBody: ResponseBody, encoding: String?, clone: Buffer): String? {
        var charset = Charset.forName("UTF-8")
        val contentType = responseBody.contentType()
        if (contentType != null) {
            charset = contentType.charset(charset)
        }
        return if (encoding != null && encoding.equals("gzip", ignoreCase = true)) {//content使用gzip压缩
            ZipUtil.decompressForGzip(clone.readByteArray(), convertCharset(charset))//解压
        } else if (encoding != null && encoding.equals("zlib", ignoreCase = true)) {//content使用zlib压缩
            ZipUtil.decompressToStringForZlib(clone.readByteArray(), convertCharset(charset))//解压
        } else {//content没有被压缩
            clone.readString(charset)
        }
    }

    @Throws(UnsupportedEncodingException::class)
    private fun parseParams(body: RequestBody, requestbuffer: Buffer): String {
        if (body.contentType() == null) return "Unknown"
        if (!body.contentType().toString().contains("multipart")) {
            var charset = Charset.forName("UTF-8")
            val contentType = body.contentType()
            if (contentType != null) {
                charset = contentType.charset(charset)
            }
            return URLDecoder.decode(requestbuffer.readString(charset), convertCharset(charset))
        }
        return "This Params isn't Text"
    }

    private fun isParseable(responseBody: ResponseBody): Boolean {
        return if (responseBody.contentLength() == 0L) false else responseBody.contentType().toString().contains("text") || isJson(responseBody)
    }

    private fun isJson(responseBody: ResponseBody): Boolean {
        return responseBody.contentType().toString().contains("json")
    }

    private fun convertCharset(charset: Charset): String {
        val s = charset.toString()
        val i = s.indexOf("[")
        return if (i == -1) s else s.substring(i + 1, s.length - 1)
    }

}